tsdb/isolation.go

```diff
@@ -0,0 +1,220 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"sync"
+)
+
+// IsolationState holds the isolation information.
+type IsolationState struct {
+	// We will ignore all writes above the max, or that are incomplete.
+	maxWriteID       uint64
+	incompleteWrites map[uint64]struct{}
+	lowWaterMark     uint64 // Lowest of incompleteWrites/maxWriteId.
+	isolation        *isolation
+
+	// Doubly linked list of active reads.
+	next *IsolationState
+	prev *IsolationState
+}
+
+// Close closes the state.
+func (i *IsolationState) Close() {
+	i.isolation.readMtx.Lock()
+	i.next.prev = i.prev
+	i.prev.next = i.next
+	i.isolation.readMtx.Unlock()
+}
+
+// isolation is the global isolation state.
+type isolation struct {
+	// Mutex for accessing writeLastId and writesOpen.
+	writeMtx sync.Mutex
+	// Each write is given an internal id.
+	lastWriteID uint64
+	// Which writes are currently in progress.
+	writesOpen map[uint64]struct{}
+	// Mutex for accessing readsOpen.
+	// If taking both writeMtx and readMtx, take writeMtx first.
+	readMtx sync.Mutex
+	// All current in use isolationStates. This is a doubly-linked list.
+	readsOpen *IsolationState
+}
+
+func newIsolation() *isolation {
+	isoState := &IsolationState{}
+	isoState.next = isoState
+	isoState.prev = isoState
+
+	return &isolation{
+		writesOpen: map[uint64]struct{}{},
+		readsOpen:  isoState,
+	}
+}
+
+// lowWatermark returns the writeId below which
+// we no longer need to track which writes were from
+// which writeId.
+func (i *isolation) lowWatermark() uint64 {
+	i.writeMtx.Lock() // Take writeMtx first.
+	defer i.writeMtx.Unlock()
+	i.readMtx.Lock()
+	defer i.readMtx.Unlock()
+	if i.readsOpen.prev == i.readsOpen {
+		return i.lastWriteID
+	}
+	return i.readsOpen.prev.lowWaterMark
+}
+
+// State returns an object used to control isolation
+// between a query and writes. Must be closed when complete.
+func (i *isolation) State() *IsolationState {
+	i.writeMtx.Lock() // Take write mutex before read mutex.
+	defer i.writeMtx.Unlock()
+	isoState := &IsolationState{
+		maxWriteID:       i.lastWriteID,
+		lowWaterMark:     i.lastWriteID,
+		incompleteWrites: make(map[uint64]struct{}, len(i.writesOpen)),
+		isolation:        i,
+	}
+	for k := range i.writesOpen {
+		isoState.incompleteWrites[k] = struct{}{}
+		if k < isoState.lowWaterMark {
+			isoState.lowWaterMark = k
+		}
+	}
+
+	i.readMtx.Lock()
+	defer i.readMtx.Unlock()
+	isoState.prev = i.readsOpen
+	isoState.next = i.readsOpen.next
+	i.readsOpen.next.prev = isoState
+	i.readsOpen.next = isoState
+	return isoState
+}
+
+// newWriteID increments the transaction counter and returns a new transaction ID.
+func (i *isolation) newWriteID() uint64 {
+	i.writeMtx.Lock()
+	i.lastWriteID++
+	writeID := i.lastWriteID
+	i.writesOpen[writeID] = struct{}{}
+	i.writeMtx.Unlock()
```

Comment:
We can use defer now without a performance penalty.

Reply:
Indeed. Done.